File: LongestValidParentheses.java
// Approach 1: Stack

class Solution {
    public int longestValidParentheses(String s) {
        Stack<Integer> stack = new Stack<>();
		int res = 0;

		for (int i = 0; i < s.length(); i++) {
			if (s.charAt(i) == '(') {
				stack.push(i);
			} else if (!stack.isEmpty() && s.charAt(stack.peek()) == '(') {  // inside valid substring, update res
				stack.pop();
				int validLen = stack.isEmpty() ? i + 1 : i - stack.peek();
				res = Math.max(res, validLen);
			} else {   // valid substring breaks
				stack.push(i);
			}
		}

		return res;
	}
}

// Approach 2: Dynamic Programming

public class Solution {
    public int longestValidParentheses(String s) {
        int[] dp = new int[s.length()];
        int result = 0;
        int leftCount = 0;
        for (int i = 0; i < s.length(); i++) {
            if (s.charAt(i) == '(') {
                leftCount++;
            } else if (leftCount > 0){
                dp[i] = dp[i - 1] + 2;
                dp[i] += (i - dp[i]) >= 0 ? dp[i - dp[i]] : 0;
                result = Math.max(result, dp[i]);
                leftCount--;
            }
        }
        return result;
    }
}

File: LongestValidParentheses.cpp
class Solution {
public:
    int longestValidParentheses(string s) {
        int n = s.size();
		
		// dp[i] denotes the length of the longest valid parentheses ending with index i
        vector<int> dp(n + 1, 0);
        stack<int> st;
        for(int i = 0; i < n; i++){
            if(st.empty()) {
                if(s[i] == '(') st.push(i);
                continue;
            }
            if(s[i] == '(') st.push(i);
            else{
                int len = i - st.top() + 1;
                st.pop();
                dp[i + 1] = len + dp[i - len + 1];
            }
        }
        int ans = 0;
        for(int i = 1; i <= n; i++) ans = max(ans, dp[i]);
        return ans;
    }
};
